mbrl/util/common.py

```diff
@@ -190,6 +194,116 @@ def create_replay_buffer(
+            be the same. Defaults to ``True``.
+        bootstrap_permutes (bool): if ``True``, the bootstrap iterator will create
+            the bootstrap data using permutations of the original data. Otherwise
+            it will use sampling with replacement. Defaults to ``False``.
+
+    Returns:
+        (tuple of :class:`mbrl.replay_buffer.TransitionIterator`): the training
+        and validation iterators, respectively.
+    """
+    data = replay_buffer.get_all(shuffle=True)
+    val_size = int(replay_buffer.num_stored * val_ratio)
+    train_size = replay_buffer.num_stored - val_size
+    train_data = data[:train_size]
+    train_iter = BootstrapIterator(
+        train_data,
+        batch_size,
+        ensemble_size,
+        shuffle_each_epoch=shuffle_each_epoch,
+        permute_indices=bootstrap_permutes,
+        rng=replay_buffer._rng,
+    )
+
+    val_iter = None
+    if val_size > 0:
+        val_data = data[train_size:]
+        val_iter = TransitionIterator(
+            val_data, batch_size, shuffle_each_epoch=False, rng=replay_buffer._rng
+        )
+
+    return train_iter, val_iter
+
+
+def get_sequence_buffer_iterator(
+    replay_buffer: ReplayBuffer,
+    batch_size: int,
+    val_ratio: float,
+    sequence_length: int,
+    ensemble_size: Optional[int] = None,
+    max_batches_per_loop: Optional[int] = None,
+) -> Tuple[SequenceTransitionIterator, Optional[SequenceTransitionIterator]]:
+    """Returns training/validation iterators for the data in the replay buffer.
+
+    Args:
+        replay_buffer (:class:`mbrl.util.ReplayBuffer`): the replay buffer from which
+            data will be sampled.
+        batch_size (int): the batch size for the iterators.
+        val_ratio (float): the proportion of data to use for validation. If 0., the
+            validation buffer will be set to ``None``.
+        sequence_length (int): the length of the sequences returned by the iterators.
+        ensemble_size (int): the number of models in the ensemble.
+        max_batches_per_loop (int, optional): if given, specifies how many batches
+            to return (at most) over a full loop of the iterator.
+
+    Returns:
+        (tuple of :class:`mbrl.replay_buffer.SequenceTransitionIterator`): the training
+        and validation iterators, respectively.
+    """
+
+    transitions = replay_buffer.get_all()
+    num_trajectories = len(replay_buffer.trajectory_indices)
+    val_size = int(num_trajectories * val_ratio)
+    train_size = num_trajectories - val_size
+    train_trajectories = replay_buffer.trajectory_indices[:train_size]
+
+    train_iterator = SequenceTransitionIterator(
+        transitions,
+        train_trajectories,
+        batch_size,
+        sequence_length,
+        ensemble_size,
+        rng=replay_buffer._rng,
+        max_batches_per_loop=max_batches_per_loop,
+    )
+
+    val_iterator = None
+    if val_size > 0:
+        val_trajectories = replay_buffer.trajectory_indices[train_size:]
```

Comment:
I think it is better to sample the trajectories for the training/validation sets. In the current implementation, the first trajectories in the buffer are used as training set and the last trajectories as validation set. If the policy changed between the rollouts that generated these trajectories, there will be a distribution mismatch between the training and validation set. 
Furthermore, this will cause that the training algorithm does not see new data until it is somewhat old (until the validation set is filled with newer data). I expect that this slows down the training of many algorithms.

Sampling the trajectories should reduce the mismatch and allow the training algorithm to get new data more quickly.

```suggestion
    trajectory_indices = random.sample(replay_buffer.trajectory_indices, len(replay_buffer.trajectory_indices))
    train_trajectories = trajectory_indices[:train_size]

    train_iterator = SequenceTransitionIterator(
        transitions,
        train_trajectories,
        batch_size,
        sequence_length,
        ensemble_size,
        rng=replay_buffer._rng,
        max_batches_per_loop=max_batches_per_loop,
    )

    val_iterator = None
    if val_size > 0:
        val_trajectories = trajectory_indices[train_size:]
```

Reply:
I agree, this was not intentional behavior, but an actual bug. There is a permutation missing that was in the original [code](https://github.com/facebookresearch/mbrl-lib/blob/master/mbrl/util/replay_buffer.py#L419). This bug also happens in the new `get_basic_buffer_iterators`. Thanks for catching!

Reply:
Fixed in https://github.com/facebookresearch/mbrl-lib/pull/91/commits/c02b702c2ce5596e2a18d950c706145e846d41f9. I realized there was no bug in `get_basic_buffer_iterators` because I  had added an option `replay_buffer.get_all(shuffle=True)`.